### PR TITLE
Karan/authentication

### PR DESCRIPTION
**Pull request by vc-vishva**
- Create post API using JWT and bcrypt packages

### CODE REVIEW

File: authentication-task/index.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+const bodyParser = require("body-parser");
+const bcrypt = require("bcrypt");
+const port = 3000;
+const jwtRouter = require("./jwt.router");
+const app = express();
+app.use("/", jwtRouter);
+
+app.use(bodyParser.json());
+
+//
```
Comment (vc-mohit) [inline authentication-task/index.js:15]: Code should be formatted & Tested properly before committing
also, remove extra lines.

File: authentication-task/jwt.router.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    // Read db.json file
+    const data = await fs.readFile("db.json");
+    const users = JSON.parse(data.toString());
+    // Check if email already exists
+    const user = users.find((u) => u.email === email);
+    if (user) {
+      return res.status(400).json({
+        error: "User already registered.",
+        user: { name, email, password: hashedPassword },
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:31]: In error response, you don't need to send whole user details, just send the error message.

File: authentication-task/jwt.router.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+      });
+    }
+    // Add new user
+    users.push({ name, email, password: hashedPassword });
+    await fs.writeFile("db.json", JSON.stringify(users));
+    res.status(200).json({
+      message: "User registered successfully.",
+      user: { name, email, password: hashedPassword },
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:39]: While sending success response, don't send a password. As of now it's ok but in live project don't do that

File: authentication-task/jwt.router.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    const isMatch = await bcrypt.compare(password, user.password);
+    if (!isMatch) {
+      return res.status(401).json({ error: "Invalid email or password." });
+    }
+    // Create and send JWT token
+    const token = jwt.sign(
+      { email: user.email, password: user.password },
+      process.env.SECRET_KEY
+    );
+    res.json({ token });
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:68]: Send JSON response with key and their value.

File: authentication-task/jwt.router.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    const response = {
+      message: "Welcome",
+      user: {
+        name: user.name,
+        email: user.email,
+      },
+    };
+
+    res.json(response);
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:110]: send JSON response only

File: authentication-task/package.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "nodemon index.js"
+  },
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "bcrypt": "^5.1.0",
+    "dotenv": "^16.0.3",
+    "epress": "^0.0.1-security",
```
Comment (vc-mohit) [inline authentication-task/package.json:14]: this package is required?

Comment (vc-vishva) [reply authentication-task/package.json:14]: bcrypt is use  for password hashing. in jwt authentication system .

File: authentication-task/jwt.router.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    // Add new user
+    users.push({ name, email, password: hashedPassword });
+    await fs.writeFile("db.json", JSON.stringify(users));
+    res.status(200).json({
+      message: "User registered successfully.",
+      user: { name, email },
+    });
+  } catch (err) {
+    res.status(500).json({ error: "Server error." });
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:35]: Send proper error messages.

File: authentication-task/jwt.router.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+jwtRouter.post("/login", async (req, res) => {
+  try {
+    const { email, password } = req.body;
+
+    const data = await fs.readFile(dbPath);
+    const users = JSON.parse(data.toString());
+    const user = users.find((u) => u.email === email);
+    if (!user) {
+      return res.status(401).json({ error: "Invalid email or password." });
```
Comment (vc-mohit) [inline authentication-task/jwt.router.js:47]: check the error messages

**Review by vc-mohit — APPROVED** [review]

**Review by vc-mohit — APPROVED** [review]

**Pull request merged**